macos/FluentUI/Button.swift

```diff
@@ -0,0 +1,207 @@
+		let opts: NSTrackingArea.Options = [.mouseEnteredAndExited, .activeAlways]
+		let trackingArea = NSTrackingArea(
+			rect: bounds,
+			options: opts,
+			owner: self,
+			userInfo: nil
+		)
+		addTrackingArea(trackingArea)
+		self.trackingArea = trackingArea
+	}
+	
+	override public func mouseEntered(with event: NSEvent) {
+		mouseEntered = true
+		needsDisplay = true
+	}
+
+	override public func mouseExited(with event: NSEvent) {
+		mouseEntered = false
+		needsDisplay = true
+	}
+	
+	override public func mouseDown(with event: NSEvent) {
+		mouseDown = true
+		needsDisplay = true
+	}
+	
+	override public func mouseUp(with event: NSEvent) {
+		mouseDown = false
+		needsDisplay = true
+	}
+	
+	open override func updateLayer() {
+		update()
+	}
+	
+	open override var intrinsicContentSize: NSSize {
+		var intrinsicContentSize = super.intrinsicContentSize;
+
+		intrinsicContentSize.width = intrinsicContentSize.width + (horizontalPadding * 2)
+		intrinsicContentSize.height += (verticalPadding * 2);
+
+		return intrinsicContentSize;
+	}
+	
+	override public var wantsUpdateLayer: Bool {
+		return true
+	}
+	
+	/// The primary color of the button, AKA, the fill color in the primaryFilled style, and the outline in the primaryOutline style
+	@objc public var primaryColor: NSColor = .clear {
+		didSet {
+            if primaryColor != oldValue {
+                update()
+            }
+		}
+	}
+
+	/// The primary style of the button
+	@objc public var style: ButtonStyle = .primaryFilled {
+		didSet {
+            if style != oldValue {
+                update()
+            }
+		}
+	}
+	
+	private func update() {
+		if let layer = layer {
+			layer.borderWidth = borderWidth
+			layer.masksToBounds = true
+			layer.cornerRadius = cornerRadius
+			layer.backgroundColor = layerBackgroundColor.cgColor
+			layer.borderColor = outlineColor.cgColor
+		}
+		self.attributedTitle = NSAttributedString(string: title, attributes: [.foregroundColor : textColor])
+	}
+	
+	private var trackingArea: NSTrackingArea?
+	
+	private var mouseEntered = false
+	private var mouseDown = false
+	
+	private var fillColor: NSColor {
+		return style == ButtonStyle.primaryFilled ? primaryColor : .clear
+	}
+	
+	private var outlineColor: NSColor {
+		return style == ButtonStyle.primaryOutline ? primaryColor.withAlphaComponent(0.4) : .clear
+	}
+	
+	private var textColor: NSColor {
+		return style == ButtonStyle.primaryFilled ? .white : primaryColor
+	}
+	
+	private var restBackgroundColor: NSColor {
+		return fillColor
+	}
+
+	private var hoverBackgroundColor: NSColor {
+		if #available(macOS 10.14, *) {
+			return fillColor.withSystemEffect(.rollover)
+		} else {
+			return fillColor.withAlphaComponent(0.5)
+		}
+	}
+
+	private var pressedBackgroundColor: NSColor {
+		if #available(macOS 10.14, *) {
+			return fillColor.withSystemEffect(.pressed)
+		} else {
+			return fillColor.withAlphaComponent(0.25)
+		}
+	}
+	
+	private var disabledBackgroundColor: NSColor {
```

Comment:
personal opinion: the outline button doesn't look disabled unless the outline color matches the title color.

Reply:
That seems fair, and the iOS / macOS buttons should align. I think that would be blocked until we get primary color shades for macOS.

Reply:
For some more context: Kaan agrees here that the outline should probably be gray, but moreso we need to align what we do with disabled colors with what Apple does. In this case, the disabled outline button has the outline color with the "disabled" system effect applied on top of it. The text in the control has some.. other filter being applied on top of it from disabling the control that I haven't quite figured out where it's coming from yet (I'm not manually greying out the text). I'll consider this not a blocker for this review